### PR TITLE
[#101] implement heredoc support

### PR DESCRIPTION
**Pull request by coryb**
Adds support for:
```
value <<EOM
  this
    is 
  literal
EOM

# undent with <<-
value <<-EOM
  this
    is
  undented
EOM

# fold to single line with <<~
value <<~EOM
  this 
  is
  folded
EOM
```

### CODE REVIEW

File: parser/cst.go [hunk 1]
```diff
@@ -24,20 +27,74 @@ var (
...
+	hereIdent lexer.Token
+}
+
+func (s *hereDocLexer) Next() (lexer.Token, error) {
+	t, err := s.lex.Next()
+	switch t.Type {
+	case s.sym["Whitespace"]:
+		if !s.keepSpace {
+			// not preserving spsace, so move on to next token
+			return s.Next()
```
Comment (hinshun) [inline parser/cst.go:79]: This is pretty elegant!

File: parser/unparse.go [hunk 2]
```diff
@@ -348,8 +365,15 @@ func (s *BlockStmt) String() string {
 			}
 		}
 
+		insideHeredoc := false
 		for i := 1; i < len(stmts); i++ {
-			if len(stmts[i]) > 0 {
+			if stmts[i] == "<HLB-RAW-HEREDOC-START>" || stmts[i] == "<HLB-RAW-HEREDOC-END>" {
```
Comment (hinshun) [inline parser/unparse.go:370]: Seems like we only need one marker? Can we move this to a constant at the top of this package. Maybe `<HLB-HEREDOC-MARKER>`.

Comment (coryb) [reply parser/unparse.go]: fixed.

**Comment by hinshun** [comment]
:rocket:

**Pull request merged**